src/mrack/providers/provider.py

```diff
@@ -174,26 +174,32 @@ async def _provision_base(self, reqs):  # pylint: disable=too-many-locals
         hosts = [self.to_host(srv) for srv in server_results]
         error_hosts = await self.parse_error_hosts(hosts)
         active_hosts = [h for h in hosts if h not in error_hosts]
-        # check ssh connectivity to succeeded hosts
-        wait_ssh = []
-        for host in active_hosts:
-            awaitable = self._wait_for_ssh(host)
-            wait_ssh.append(awaitable)
-
-        ssh_results = await asyncio.gather(*wait_ssh)
-        # We distinguish the success hosts and new error hosts from active by using:
-        # res[RET_CODE] 0 - the result of operation returned from self._wait_for_ssh()
-        # res[HOST_OBJ] 1 - the host object returned from self._wait_for_ssh()
         success_hosts = []
-        for res in ssh_results:
-            if res[RET_CODE]:
-                success_hosts.append(res[HOST_OBJ])
-            else:
-                res[HOST_OBJ].error = (
-                    "Could not establish ssh connection to host "
-                    f"{res[HOST_OBJ].host_id} with IP {res[HOST_OBJ].ip_addr}"
-                )
-                error_hosts.append(res[HOST_OBJ])
+
+        if global_context["config"].get("post_provisioning_ssh_check", True):
```

Comment:
I'm not really fan of this. IMHO global config for SSH check should be used in a very rare circumstances.  Providing this only as a global option will lead to disabling it essentially for everything even if it was used only for a certain use case - specific image.  Making ssh_check feature useless.  

Better would be to have it configured per-host with possibility to define on host level, metadata level and provisioning config level. In provisioning config, it might be also a possibility to do it per-image (os).  That way we might have a possibility to disable it only in cases where it is needed. 

I.e. I'd move the logic to transformer.

Reply:
When thinking about it, introducing global context in providers is also a path to hell. It removes separation of concerns, in long run making it a spaghetti code where everything depends on everything.

Reply:
I understand your concern about depending on global context.
I will add a parameter to the method call and fallback to global config if parameter is not set.

Reply:
I don't understand much the steps you are going to do. So will briefly re-cap: this logic should be in a tranformer as it has access to cofiguration which is provided to it.